dodiscover/constraint/pcalg.py

```diff
@@ -0,0 +1,309 @@
+import logging
+from itertools import combinations, permutations
+from typing import Dict, List, Optional, Set
+
+import networkx as nx
+
+from dodiscover.ci.base import BaseConditionalIndependenceTest
+from dodiscover.constraint.utils import is_in_sep_set
+from dodiscover.typing import Column
+
+from .._protocol import EquivalenceClassProtocol
+from ._classes import BaseConstraintDiscovery
+
+logger = logging.getLogger()
+
+
+class PC(BaseConstraintDiscovery):
+    """Peter and Clarke (PC) algorithm for causal discovery.
+
+    Assumes causal sufficiency, that is, all confounders in the
+    causal graph are observed variables. See :footcite:`Spirtes1993` for
+    full details on the algorithm.
+
+    Parameters
+    ----------
+    ci_estimator : Callable
```

Comment:
Does the fact that this is a callable leave us open to new methods in CI estimation?
If so, then great.
I think the ability to write custom CI tests using new methods is a powerful way to utilize cutting-edge research.

Reply:
Yep! It is any Callable with the `BaseConditionalIndependenceTest` API. However, I will end up refactoring all those to just "functions" as suggested by @bloebp and @petergtz. In a downstream PR this will look like the following:

- ctrl+f / replace `BaseConditionalIndependenceTest` with `Callable[[np.ndarray, np.ndarray, np.ndarray], float]` for type hints
- just turn all the CI test classes into their corresponding functions
- migrate any internal functionality of the CI test classes into private util functions for the `dodiscover/ci/` submodule

Reply:
Basically this will be very powerful because as long as CI tests adhere to a specific input/output function signature, then yes this enables anyone to arbitrarily pass in whatever CI test they want. This is an API pro over say passing in "strings", where you only allow certain CI tests.